enterprise/server/remote_execution/nbd/nbdclient/nbdclient.go

```diff
@@ -0,0 +1,299 @@
+	if d.mountPath != "" {
+		return status.FailedPreconditionError("already mounted")
+	}
+	if err := d.createDevice(); err != nil {
+		return status.WrapError(err, "failed to create device")
+	}
+	fstype := filesystemTypeString(d.metadata.GetFilesystemType())
+	if err := syscall.Mount(d.DevicePath(), path, fstype, syscall.MS_RELATIME, "" /*=data*/); err != nil {
+		return status.InternalErrorf("failed to mount %s (type %s) to %s: %s", d.DevicePath(), fstype, path, err)
+	}
+	d.mountPath = path
+	log.Infof("Mounted %s (type %s) to %s", d.DevicePath(), d.metadata.GetName(), path)
+	return nil
+}
+
+// Unmount unmounts the device from its current mount path and disconnects the
+// device.
+func (d *ClientDevice) Unmount() error {
+	if d.mountPath == "" {
+		return status.FailedPreconditionError("not mounted")
+	}
+	if err := syscall.Unmount(d.mountPath, 0); err != nil {
+		return status.InternalErrorf("unmount %s (%s): %s", d.mountPath, d.metadata.GetName(), err)
+	}
+	log.Infof("Unmounted %s", d.mountPath)
+	d.mountPath = ""
+	// Also disconnect the device since the host may be swapping out the backing
+	// file, and it's simpler to disconnect and reconnect rather than try to
+	// reconfigure a connected device.
+	if err := d.device.Disconnect(); err != nil {
+		return status.InternalErrorf("disconnect %s (%s): %s", d.DevicePath(), d.metadata.GetName(), err)
+	}
+	log.Infof("Disconnected %s", d.DevicePath())
+	d.device = nil
+	return nil
+}
+
+func (d *ClientDevice) ReadAt(p []byte, off int64) (int, error) {
+	var res *nbdpb.ReadResponse
+	err := d.doWithRedial(func() error {
+		var err error
+		res, err = d.client.Read(d.ctx, &nbdpb.ReadRequest{
+			Name:   d.metadata.GetName(),
+			Offset: off,
+			Length: int64(len(p)),
+		})
+		return err
+	})
+	if err != nil {
+		fatalf("%s: read failed: %s", d.DevicePath(), err)
+	}
+	if len(res.Data) > len(p) {
+		fatalf(
+			"block device server returned too many bytes (0x%x) for read at offset=0x%x length=0x%x",
+			len(res.Data), len(p), off)
+	}
+	copy(p, res.Data)
+	return len(res.Data), nil
+}
+
+func (d *ClientDevice) WriteAt(p []byte, off int64) (int, error) {
+	err := d.doWithRedial(func() error {
+		_, err := d.client.Write(d.ctx, &nbdpb.WriteRequest{
+			Name:   d.metadata.GetName(),
+			Offset: off,
+			Data:   p,
+		})
+		return err
+	})
+	if err != nil {
+		fatalf("%s: write failed: %s", d.DevicePath(), err)
+	}
+	return len(p), nil
+}
+
+func (d *ClientDevice) Sync() error {
+	// Do nothing for now since we don't buffer reads/writes and instead pass
+	// them through directly to the host executor. The executor is responsible
+	// for guaranteeing that reads and writes are persisted.
+	return nil
+}
+
+// doWithRedial invokes a function that performs an RPC to the host server,
+// and re-dials the host server if the function fails on the first attempt.
+func (d *ClientDevice) doWithRedial(f func() error) error {
+	var lastErr error
+	for i := 1; i <= 2; i++ {
```

Comment:
I wanted the retry to be immediate (no delay) and only happen once, since we're just reconnecting to the new server which should be guaranteed to be running. It seemed more straightforward to use a loop than to use a retry.